src/in_process_server.cpp

```diff
@@ -35,6 +35,407 @@
 #include <algorithm>
 #include <experimental/optional>
 #include <unordered_map>
+#include <wayland-server-core.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <mutex>
+#include <thread>
+
+namespace
+{
+template<typename Callable>
+struct callable_traits : public callable_traits<decltype(&Callable::operator())> {};
+
+template<typename Callable, typename Returns, typename... Args>
+struct callable_traits<Returns(Callable::*)(Args...) const>
+{
+    using return_type = Returns;
+    using args = typename std::tuple<Args...>;
+};
+
+/*
+ * call_helper and call are downmarket versions of C++17's std::invoke()
+ *
+ * We don't yet require C++17, so downmarket it is!
```

Comment:
Maybe we should? I don't think we would have a problem on any supported OS.

Reply:
Although... on 16.04LTS, g++5.4 lacks `invoke()` from `<functional>` despite accepting `-std=c++17`.